hbase-server/src/main/java/org/apache/hadoop/hbase/master/MetaFixer.java

```diff
@@ -86,74 +87,176 @@ void fix() throws IOException {
+
     if (left.getTable().equals(right.getTable())) {
       // Simple case.
       if (Bytes.compareTo(left.getEndKey(), right.getStartKey()) >= 0) {
-        LOG.warn("Skipping hole fix; left-side endKey is not less than right-side startKey; " +
-            "left=<{}>, right=<{}>", left, right);
-        return holeCover;
-      }
-      holeCover = buildRegionInfo(left.getTable(), left.getEndKey(), right.getStartKey());
-    } else {
-      boolean leftUndefined = left.equals(RegionInfo.UNDEFINED);
-      boolean rightUnefined = right.equals(RegionInfo.UNDEFINED);
-      boolean last = left.isLast();
-      boolean first = right.isFirst();
-      if (leftUndefined && rightUnefined) {
-        LOG.warn("Skipping hole fix; both the hole left-side and right-side RegionInfos are " +
-            "UNDEFINED; left=<{}>, right=<{}>", left, right);
-        return holeCover;
-      }
-      if (leftUndefined || last) {
-        holeCover = buildRegionInfo(right.getTable(), HConstants.EMPTY_START_ROW,
-            right.getStartKey());
-      } else if (rightUnefined || first) {
-        holeCover = buildRegionInfo(left.getTable(), left.getEndKey(), HConstants.EMPTY_END_ROW);
-      } else {
-        LOG.warn("Skipping hole fix; don't know what to do with left=<{}>, right=<{}>",
-            left, right);
-        return holeCover;
+        LOG.warn("Skipping hole fix; left-side endKey is not less than right-side startKey;"
+          + " left=<{}>, right=<{}>", left, right);
+        return Optional.empty();
       }
+      return Optional.of(buildRegionInfo(left.getTable(), left.getEndKey(), right.getStartKey()));
     }
-    return holeCover;
+
+    final boolean leftUndefined = left.equals(RegionInfo.UNDEFINED);
+    final boolean rightUndefined = right.equals(RegionInfo.UNDEFINED);
+    final boolean last = left.isLast();
+    final boolean first = right.isFirst();
+    if (leftUndefined && rightUndefined) {
+      LOG.warn("Skipping hole fix; both the hole left-side and right-side RegionInfos are " +
+        "UNDEFINED; left=<{}>, right=<{}>", left, right);
+      return Optional.empty();
+    }
+    if (leftUndefined || last) {
+      return Optional.of(
+        buildRegionInfo(right.getTable(), HConstants.EMPTY_START_ROW, right.getStartKey()));
+    }
+    if (rightUndefined || first) {
+      return Optional.of(
+        buildRegionInfo(left.getTable(), left.getEndKey(), HConstants.EMPTY_END_ROW));
+    }
+    LOG.warn("Skipping hole fix; don't know what to do with left=<{}>, right=<{}>", left, right);
+    return Optional.empty();
   }
 
-  private RegionInfo buildRegionInfo(TableName tn, byte [] start, byte [] end) {
+  private static RegionInfo buildRegionInfo(TableName tn, byte [] start, byte [] end) {
     return RegionInfoBuilder.newBuilder(tn).setStartKey(start).setEndKey(end).build();
   }
 
+  /**
+   * Create entries in the filesystem for each provided {@link RegionInfo}. Best effort.
```

Comment:
We have to do this?  Ain't they auto-created on region open?

Reply:
We have 9 method variants for opening a region...

Ah yes. Looks like it's done through the `Region#initialize` method, called as part of `HRegion#open`.